### PR TITLE
Don't rely on entry point to be npm-cli.js

### PR DESCRIPTION
**Pull request by olydis**
fixes https://github.com/npm/npm/issues/18449

### CODE REVIEW

File: lib/npm.js [hunk 1]
```diff
@@ -70,8 +70,7 @@
 
   try {
     // startup, ok to do this synchronously
-    var j = parseJSON(fs.readFileSync(
-      path.join(__dirname, '../package.json')) + '')
+    var j = require(path.join(__dirname, '../package.json'))
```
Comment (iarna) [inline lib/npm.js]: We don't use require for a reason (it provides poor error messages when there are errors in the file and I don't think it supports files starting with UTF-8 BOMs.

File: lib/pack.js [hunk 2]
```diff
@@ -171,7 +171,7 @@ function packGitDep (manifest, dir) {
         return acc
       }, [])
       const child = cp.spawn(process.env.NODE || process.execPath, [
-        require.main.filename,
+        require.resolve('../bin/npm-cli.js'),
```
Comment (iarna) [inline lib/pack.js:174]: This seems reasonable

**Review by iarna — CHANGES_REQUESTED** [review]
So the thing is:

We don't support using npm as a package. We can and do change APIs related to that without ticking the major version number. That it works as well as it does for you is more of an accident than a fature.

That said: I'm not opposed changing `require.main.filename` to `npm-cli.js`. I'm ok with taking this PR, but only without the change to how the `package.json` is required.

File: lib/npm.js (hunk 1 above)
Comment (olydis) [reply lib/npm.js]: > I don't think it supports files starting with UTF-8 BOMs.

it does

Comment (olydis) [reply lib/npm.js]: but this change is not at all important, the error message part makes perfect sense :) will undo

**Comment by olydis** [comment]
@iarna Thanks, what you say makes perfect sense! I have reverted the `require` part. We call `npm` functions at a very "close to CLI" level, but are seriously considering the switch to child_process as soon as we get to it. Rest assured, we don't and won't expect the slightest support for what we did 😄

**Review by zkat — APPROVED** [review]
I'm ok with this change. Thanks!

**Pull request merged**